### PR TITLE
WIP: Variable editor for object groups

### PR DESCRIPTION
**Pull request by D8H**

### CODE REVIEW

File: Core/GDCore/IDE/Events/EventsVariableInstructionTypeSwitcher.h [hunk 1]
```diff
@@ -46,8 +55,13 @@ class GD_CORE_API EventsVariableInstructionTypeSwitcher
 
   const gd::Platform &platform;
   const gd::VariablesContainer &targetVariablesContainer;
-  gd::String objectName;
+  /**
+   * Groups don't have VariablesContainer
+   */
+  const gd::String groupName;
```
Comment (4ian) [inline Core/GDCore/IDE/Events/EventsVariableInstructionTypeSwitcher.h:61]: This states a fact but does not explain the reason for the variable. Be more precise:
```suggestion
   * Groups don't have VariablesContainer, so [targetVariablesContainer will be pointing to nullVariablesContainer and we use a group name instead to check for....??]
   */
  const gd::String groupName;
```

File: Core/GDCore/IDE/Events/EventsVariableReplacer.cpp [hunk 2]
```diff
@@ -197,10 +205,11 @@ class GD_CORE_API ExpressionVariableReplacer
     // (and if it's a variable reference or a value does not have any importance
     // here).
 
-    if (forcedInitialVariablesContainer) {
+    if (forcedVariablesContainer) {
       // A scope was forced. Honor it: it means this node represents a variable
       // of the forced variables container.
-      if (forcedInitialVariablesContainer == &targetVariablesContainer) {
+      if (forcedVariablesContainer == &targetVariablesContainer ||
+          (forcedObjectName == targetGroupName && !targetGroupName.empty())) {
```
Comment (4ian) [inline Core/GDCore/IDE/Events/EventsVariableReplacer.cpp:212]: All of these `xxx == targetGroupName && !targetGroupName.empty()` would almost mandate a `IsTheTargetGroupName(xxx)` helper function.

For the reader, it's hard to parse (especially because you check for targetGroupName not empty at the end, which is counter intuitive in my opinion). Let's make a helper so that it's easier to read the logic.

File: Core/GDCore/IDE/Events/EventsVariableReplacer.cpp [hunk 3]
```diff
@@ -382,10 +394,15 @@ class GD_CORE_API ExpressionVariableReplacer
...
   const gd::Platform& platform;
   const gd::ProjectScopedContainers& projectScopedContainers;
-  const gd::VariablesContainer* forcedInitialVariablesContainer;
+  const gd::VariablesContainer* forcedVariablesContainer;
+  gd::String forcedObjectName;
 
   // Renaming or removing to do:
   const gd::VariablesContainer& targetVariablesContainer;
+  /**
+   * Groups don't have VariablesContainer
```
Comment (4ian) [inline Core/GDCore/IDE/Events/EventsVariableReplacer.cpp:403]: Ditto as in the other file

File: Core/GDCore/IDE/Events/EventsVariableReplacer.h [hunk 4]
```diff
@@ -55,9 +66,14 @@ class GD_CORE_API EventsVariableReplacer
 
   const gd::Platform &platform;
   const gd::VariablesContainer &targetVariablesContainer;
-  gd::String objectName;
+  /**
+   * Groups don't have VariablesContainer
```
Comment (4ian) [inline Core/GDCore/IDE/Events/EventsVariableReplacer.h:70]: Same here

File: Core/GDCore/IDE/GroupVariableHelper.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Help handling variables of group objects as a whole.
+ *
+ * This is used by the object group variable editor.
+ */
+class GD_CORE_API GroupVariableHelper {
+public:
+  /**
+   * Copy every variable from every object of the group to the other object
```
Comment (4ian) [inline Core/GDCore/IDE/GroupVariableHelper.h:28]:
```suggestion
   * Copy every variable from every object of the group to the other objects
```

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp [hunk 6]
```diff
@@ -140,6 +141,7 @@ void WholeProjectRefactorer::EnsureObjectEventsFunctionsProperParameters(
   }
 }
 
+// TODO Handle position changes for group variables.
```
Comment (4ian) [inline Core/GDCore/IDE/WholeProjectRefactorer.cpp:144]: Can you be more precise on what this means?

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp [hunk 7]
```diff
@@ -321,8 +330,83 @@ void WholeProjectRefactorer::ApplyRefactoringForVariablesContainer(
...
+    const gd::VariablesChangeset &changeset,
+    const gd::SerializerElement &originalSerializedVariables) {
+
+  // While we support refactoring that would remove all references (actions, conditions...)
+  // it's both a bit dangerous for the user and we would need to show the user what
+  // will be removed before doing so. For now, just clear the removed variables so they don't
+  // trigger any refactoring.
+  std::unordered_set<gd::String> removedVariableNames;
+
+  // Rename variables in events for the group objects.
```
Comment (4ian) [inline Core/GDCore/IDE/WholeProjectRefactorer.cpp:353]:
```suggestion
  // Rename variables in events for the objects group.
```

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp [hunk 8]
```diff
@@ -321,8 +330,83 @@ void WholeProjectRefactorer::ApplyRefactoringForVariablesContainer(
...
+      objectGroup, changeset);
+
+  // Switch types of instructions for the group objects.
+  for (const gd::String &objectName : objectGroup.GetAllObjectsNames()) {
+    const bool hasObject = objectsContainer.HasObjectNamed(objectName);
+    if (!hasObject && !globalObjectsContainer.HasObjectNamed(objectName)) {
+      continue;
+    }
+    auto &object = hasObject ? objectsContainer.GetObject(objectName)
+                             : globalObjectsContainer.GetObject(objectName);
```
Comment (4ian) [inline Core/GDCore/IDE/WholeProjectRefactorer.cpp:394]: Could almost be a helper function (with a comment that it should probably one day be an ObjectsContainersList. Passing two ObjectsContainer is a smell that it's not future proof).

File: Core/GDCore/Project/Variable.cpp [hunk 9]
```diff
@@ -411,4 +425,64 @@ void Variable::CopyChildren(const gd::Variable& other) {
...
+  }
+  if (type == Variable::Type::String) {
+    return str == variable.str;
+  }
+  if (type == Variable::Type::Boolean) {
+    return boolVal == variable.boolVal;
+  }
+  if (type == Variable::Type::Boolean) {
+    return boolVal == variable.boolVal;
+  }
```
Comment (4ian) [inline Core/GDCore/Project/Variable.cpp:450]:
```suggestion
  if (type == Variable::Type::Boolean) {
    return boolVal == variable.boolVal;
  }

```

File: newIDE/app/src/EventsSheet/ParameterFields/ObjectVariableField.js [hunk 10]
```diff
@@ -94,6 +95,14 @@ export default React.forwardRef<ParameterFieldProps, ParameterFieldInterface>(
...
       parameterIndex,
     });
+    const objectGroup = objectName
+      ? getObjectGroupByName(
+          globalObjectsContainer,
+          objectsContainer,
+          objectName
+        )
+      : null;
+    console.log(objectGroup);
```
Comment (4ian) [inline newIDE/app/src/EventsSheet/ParameterFields/ObjectVariableField.js:105]: To remove

File: newIDE/app/src/stories/componentStories/LayoutEditor/ObjectGroupEditorDialog.stories.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import * as React from 'react';
+import { action } from '@storybook/addon-actions';
+
+// Keep first as it creates the `global.gd` object:
+import { testProject } from '../../GDevelopJsInitializerDecorator';
+
+import ObjectGroupEditorDialog from '../../../ObjectGroupEditor/ObjectGroupEditorDialog';
+
+export default {
+  title: 'LayoutEditor/ObjectGroupEditor',
```
Comment (4ian) [inline newIDE/app/src/stories/componentStories/LayoutEditor/ObjectGroupEditorDialog.stories.js:12]:
```suggestion
  title: 'LayoutEditor/ObjectGroupEditorDialog',
```

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+          <VariablesList
+            projectScopedContainersAccessor={projectScopedContainersAccessor}
+            variablesContainer={groupVariablesContainer.current}
+            areObjectVariables
+            emptyPlaceholderTitle={
+              <Trans>Add your first object group variable</Trans>
+            }
+            emptyPlaceholderDescription={
+              <Trans>
+                These variables hold additional information on an object group.
```
Comment (4ian) [inline newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:197]:
```suggestion
                These variables hold additional information and are available on all objects of the group.
```

File: newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+          areObjectVariables
+          initiallySelectedVariableName={
+            actualInitiallySelectedVariableName.current
+          }
+          emptyPlaceholderTitle={
+            <Trans>Add your first object group variable</Trans>
+          }
+          emptyPlaceholderDescription={
+            <Trans>
+              These variables hold additional information on an object group.
```
Comment (4ian) [inline newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js:198]:
```suggestion
              These variables hold additional information and are available on all objects of the group.
```

File: newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+import HelpButton from '../UI/HelpButton';
+import { getVariableContextFromNodeId } from './VariableToTreeNodeHandling';
+import { ProjectScopedContainersAccessor } from '../InstructionOrExpression/EventsScope.flow';
+import { insertInVariablesContainer } from '../Utils/VariablesUtils';
+import { getRootVariableName } from '../EventsSheet/ParameterFields/VariableField';
+import { getNodeIdFromVariableName } from './VariableToTreeNodeHandling';
+
+const gd: libGDevelop = global.gd;
+
+const getVariablePathFromNodeId = (
```
Comment (4ian) [inline newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js:22]: This is copy pasted from somewhere else?

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+      group,
+      changeset,
+      originalSerializedVariables
+    );
+    groupVariablesContainer.current.clearPersistentUuid();
+  };
+
+  const [groupObjectNames, setGroupObjectNames] = React.useState<Array<string>>(
+    () => group.getAllObjectsNames().toJSArray()
+  );
```
Comment (4ian) [inline newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:97]: Is this worth duplicating this information in the state? Any duplication runs the risk of being desynchronized if someones forget to call setGroupObjectNames.

It might be slightly less efficient, but more bug-prone, to just use `group.getAllObjectsNames().toJSArray()` when you need it.

If it's an optimisation, then rename it so that it's obvious for the reader `cachedGroupObjectNames` and add a comment explaining why "// Cache the object names to avoid recomputing them at every render".

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp (hunk 7 above)
Comment (D8H) [reply Core/GDCore/IDE/WholeProjectRefactorer.cpp:353]: Actually, I meant "the objects of the group".

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp (hunk 8 above)
Comment (D8H) [reply Core/GDCore/IDE/WholeProjectRefactorer.cpp:394]: I agree, but `ObjectsContainersList` only grant read-only access. There is maybe a distinction to do between moving the context and doing changes in the context.
Maybe moving the context should be done by creating a new context. A `const ObjectsContainersList` would mean that no changes can be done to the objects.

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    onCancel,
+  });
+
+  const [currentTab, setCurrentTab] = React.useState<ObjectGroupEditorTab>(
+    initialTab || 'objects'
+  );
+
+  const groupVariablesContainer = ((React.useRef<gdVariablesContainer | null>(
+    null
+  ): any): { current: gdVariablesContainer });
```
[the remark this answers is not in the document]
Comment (AlexandreSi) [reply newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:57]: And with what I see below, you could write:
```js
const groupVariablesContainer = React.useRef<?gdVariablesContainer>(
  gd.GroupVariableHelper.mergeVariableContainers(
      projectScopedContainersAccessor.get().getObjectsContainersList(),
      group
    ) 
)
```

File: newIDE/app/src/ObjectGroupEditor/NewObjectGroupEditorDialog.js [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+      onRequestClose={onCancel}
+      onApply={apply}
+      open
+      maxWidth="sm"
+      fixedContent={
+        <ColumnStackLayout>
+          {!isGroupAlreadyAdded && (
+            <SemiControlledTextField
+              fullWidth
+              id="object-name"
```
Comment (AlexandreSi) [inline newIDE/app/src/ObjectGroupEditor/NewObjectGroupEditorDialog.js:108]: id to update

File: newIDE/app/src/ObjectGroupsList/ObjectGroupsListWithObjectGroupEditor.js [hunk 18]
```diff
@@ -26,68 +28,81 @@ type Props = {|
   unsavedChanges?: ?UnsavedChanges,
 |};
 
-type State = {|
```
Comment (AlexandreSi) [inline newIDE/app/src/ObjectGroupsList/ObjectGroupsListWithObjectGroupEditor.js:29]: 🤌

File: newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+          emptyPlaceholderTitle={
+            <Trans>Add your first object group variable</Trans>
+          }
+          emptyPlaceholderDescription={
+            <Trans>
+              These variables hold additional information and are available on
+              all objects of the group.
+            </Trans>
+          }
+          helpPagePath={'/all-features/variables/object-variables'}
```
Comment (AlexandreSi) [inline newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js:191]: Should this be updated?

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (D8H) [reply newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:57]: This is to avoid to check nullity everywhere when we know it will never be null because of the following if.

File: newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js (hunk 19 above)
Comment (D8H) [reply newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js:191]: I think the link should be kept.
I guess that we'll add a note in the `object-variables` pages about "mixed values" and link to a section in the object group page that explains what happens to objects when variables are modified in the group editor (which will link back to the `object-variables` page).

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (D8H) [reply newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:57]: Sorry, I missed your last message.
This is what I did at first but it called `mergeVariableContainers` at every update so it was resetting the variables when the tab changed.

[the remark this answers is not in the document]
Comment (D8H) [reply newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:57]: The ref is not reinitialized, but the parameters are evaluated so the function is called and as it returns a `VariablesContainer` by value, it reuse the same instance as the `current`.

File: Core/GDCore/IDE/WholeProjectRefactorer.cpp (hunk 8 above)
Comment (4ian) [reply Core/GDCore/IDE/WholeProjectRefactorer.cpp:393]: Ok, we'll see later how we pass "not just read only" list of objects containers :)

File: newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (4ian) [reply newIDE/app/src/ObjectGroupEditor/EditedObjectGroupEditorDialog.js:60]: I think I found the hooks we made for this :)

useValueWithInit
useRefWithInit

Try one of these

File: newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+  objectGroup,
+  onCancel,
+  onApply,
+  open,
+  hotReloadPreviewButtonProps,
+  initiallySelectedVariableName,
+  shouldCreateInitiallySelectedVariable,
+  onComputeAllVariableNames,
+}: Props) => {
+  const groupVariablesContainer = ((React.useRef<gdVariablesContainer | null>(
```
[the remark this answers is not in the document]
Comment (4ian) [reply newIDE/app/src/VariablesList/ObjectGroupVariablesDialog.js:53]: useRefWithInit (or useValueWithInit) should work here too

**Review by 4ian — APPROVED** [review]
Ok once the useRefWithInit (or useValueWithInit) is used + help link updated.
